### DISCUSSION TITLE
GraphQL controllers and policies called only for top level objects

### DISCUSSION THREAD

**Discussion by abanchev** [strapi/strapi#6108] [Legacy Topics]
**Describe the bug**

I have three collections A, B, and C.
When i request nested data using graphql the **controller functions do not get called** for each sub object in the path that we request.
**Neither do the policies or resolvers** for B and C if the root object is A.
And what is more the controller for A may see and modify some of the fields for B, however they get populated at a later time again. And the A controller never sees the data in C.

**Steps to reproduce the behavior**
Check the example repo

**Expected behavior**
To be able to inspect/modify the data using the controllers that are setup for the different collections

**Code snippets**
Here is a repo that illustrates the problem
https://github.com/abanchev/strapi-issues-test

**System**

- Node.js version: 12
- NPM version: 6.14.4
- Strapi version: v3.0.0-beta.20.3
- Database: mysql
- Operating system: linux

**Additional context**
The issue was discussed here:
https://github.com/strapi/strapi/discussions/5875

**Comment by lauriejim** [comment]
Ping @alexandrebodin about that.
What we plan to do about this topic.

**Comment by alexandrebodin** [comment]
Hi, Right now this is working as expected. Only the top level resolver calls the controller. Calling controller for nested fields is not something that we want or should do as a controller really is supposed to represent a top level access (API params, filtering and such) When resolving nested fields we need better absraction than controllers. More generally we should shift to using services and it might make more sense to use services to fetch nested fields in that case.

Last topic I want to mention is that nested resolvers use dataloader extensively to try and optimize performances. Calling either services or controllers would remove this feature if we don't think it through. 

Right now we don't plan on changing the behaviour in the short term

**Reply by abanchev** [reply]
How should we handle filtering in this case? As the data is populated after the controller gets executed.
